### PR TITLE
Refresh

### PR DESCRIPTION
**Pull request by rieschl**
* require PHP 7.4
* adhere to new coding standard (PSR-12)
* update PHPUnit
* add code quality tools (PHPStan, Infection)
* normalize composer.json
* add GitHub CI

### CODE REVIEW

**Comment by rieschl** [comment]
CI: https://github.com/rieschl/automatic-di/runs/1400901573?check_suite_focus=true

PHP8 support will be added after this PR is merged.

**Comment by rieschl** [comment]
`\Interop\Container\ContainerInterface` should be replaced by `\Psr\Container\ContainerInterface`. Should I do that in this PR? It should not cause a BC break, because `InteropContainer` extends `PsrContainer`.

File: src/AutomaticDiConfig.php [hunk 1]
```diff
@@ -1,43 +1,66 @@
...
+
 class AutomaticDiConfig
 {
-    /** @var array */
-    private $preferences;
-    /** @var array */
-    private $classPreferences;
+    /** @var array<string, string> */
+    private array $preferences;
+    /** @var array<string, array<string, string>> */
```
Comment (MidnightDesign) [inline src/AutomaticDiConfig.php]: You should use the more specific `class-string` here.

**Comment by MidnightDesign** [comment]
> `\Interop\Container\ContainerInterface` should be replaced by `\Psr\Container\ContainerInterface`. Should I do that in this PR? It should not cause a BC break, because `InteropContainer` extends `PsrContainer`.

You can include that in this PR.

**Review by MidnightDesign — APPROVED** [review]

File: src/AutomaticDiConfig.php [hunk 2]
```diff
@@ -1,43 +1,69 @@
...
+            // @phpstan-ignore-next-line
             if (!array_key_exists($key, $config) || !is_array($config[$key])) {
                 throw new InvalidArgumentException(sprintf('Missing or invalid config key "%s".', $key));
             }
         }
     }
 
-    public function getPreferences():array
+    /**
+     * @return array<string, string>
```
Comment (MidnightDesign) [inline src/AutomaticDiConfig.php]:
```suggestion
     * @return array<class-string, class-string>
```

File: src/AutomaticDiContainer.php [hunk 3]
```diff
@@ -1,47 +1,61 @@
...
 
     public function __construct(ContainerInterface $container, AutomaticDiConfig $config)
     {
         $this->container = $container;
         $this->config = $config;
     }
 
+    /**
+     * @param string $id
+     * @return mixed|object
```
Comment (MidnightDesign) [inline src/AutomaticDiContainer.php]: `mixed` is a supertype of `object`.
```suggestion
     * @return mixed
```

File: src/AutomaticDiContainer.php [hunk 4]
```diff
@@ -1,47 +1,61 @@
...
-        if (interface_exists($id) && $this->hasPreference($id)) {
-            return true;
-        }
-        return false;
+        return interface_exists($id) && $this->hasPreference($id);
     }
 
-    private function createConstructorArgs(ReflectionClass $reflectionClass):array
+    /**
+     * @return array<int, mixed>
```
Comment (MidnightDesign) [inline src/AutomaticDiContainer.php]:
```suggestion
     * @return list<mixed>
```

File: src/AutomaticDiContainer.php [hunk 5]
```diff
@@ -98,12 +115,15 @@ private function getPreference(string $className):string
         return $className;
     }
 
-    private function getPreferences():array
+    /**
+     * @return array<string, string>
```
Comment (MidnightDesign) [inline src/AutomaticDiContainer.php]:
```suggestion
     * @return array<class-string, class-string>
```

File: tests/unit/AutomaticDiConfigTest.php [hunk 6]
```diff
@@ -1,40 +1,48 @@
...
-        $this->assertArrayHasKey('foo', $classPreferences[Baz::class]);
-        $this->assertSame(Foo::class, $classPreferences[Baz::class]['foo']);
+        self::assertArrayHasKey(Baz::class, $classPreferences);
+        self::assertArrayHasKey('foo', $classPreferences[Baz::class]);
+        self::assertSame(Foo::class, $classPreferences[Baz::class]['foo']);
     }
 
-    public function invalidConfigData()
+    /**
+     * @return array<array<array<string, mixed>>>
```
Comment (MidnightDesign) [inline tests/unit/AutomaticDiConfigTest.php]:
```suggestion
     * @return list<array{array<string, mixed>}>
```

**Review by MidnightDesign — APPROVED** [review]

**Pull request merged**